doc/ccze.1.in
.\" Hey Emacs! This is -*- nroff -*-!
.\" Copyright (C) 2002, 2003 Gergely Nagy <algernon@@bonehunter.rulez.org>
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one.
.\"
.\" Permission is granted to copy and distribute translations of this
.\" manual into another language, under the above conditions for modified
.\" versions, except that this permission notice may be stated in a
.\" translation approved by the Author.
.TH CCZE 1 "2003-04-24" "CCZE @VERSION@" CCZE
.SH NAME
ccze \- A robust log colorizer
.SH SYNOPSIS
.B ccze
.RI [ options "] <" logfile
.SH DESCRIPTION
This manual page documents briefly the
.B ccze
utility, which is a drop\-in replacement for
.BR colorize ,
but written in C, to be faster and less resource\-hungry. The goal was
to be fully backwards compatible, yet superior with respect to speed
and features.
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`\-'). A summary of options is
included below.
.TP
.B \-a, \-\-argument \fIPLUGIN=ARGUMENTS\fB
Use this option to pass \fIARGUMENTS\fR to the specified
\fIPLUGIN\fR. The argument list is whitespace separated.
.TP
.B \-A, \-\-raw\-ansi
If one wants to enable raw ANSI color sequences instead of using
curses, this option does just that.
.TP
.B \-c, \-\-color \fIKEY=COLOR\fB
Set the color of the keyword \fIKEY\fR to \fICOLOR\fR, like one would
do in one of the configuration files.
.TP
.B \-C, \-\-convert\-date
Convert unix timestamp to readable date format (in
oops and squid logs, for example).
.TP
.B \-F, \-\-rcfile \fIrcfile\fB
Read \fIrcfile\fR as a configuration file upon startup, instead of the
default ones.
.TP
.B \-h, \-\-html
Instead of colorising the input onto the console, output it in HTML
format instead.
.TP
.B \-l, \-\-list\-plugins
List all available (loaded) plugins, along with their type and a short
description.
.TP
.B \-m, \-\-mode \fImode\fB
Change the output mode. Available modes are \fIcurses\fR, \fIansi\fR
and \fIhtml\fR.
.TP
.B \-o, \-\-options \fIOPTIONS...\fB
\fBCCZE\fR is able to toggle some of its features with this
option. You can toggle the \fIscroll\fR, \fIwordcolor\fR,
\fIlookups\fR, and \fItransparent\fR features, or you can fiddle with
\fIcssfile\fR. All of these are enabled by default, except
\fIcssfile\fR. One can turn them off by prefixing the option with a
"no".

With \fIscroll\fR, one can enable or disable scrolling. If the output
is not redirected, it is wise to leave it enabled.

The \fIwordcolor\fR option makes \fBccze\fR search for different
keywords in unparsed input, and color those too. Since it is quite
fast, and makes the output look better, it is recommended to leave
this enabled.

However, \fIlookups\fR is an option that might be better to
disable. When on, \fBccze\fR will try to look up usernames, service
names, and lots of other stuff, which will slow down coloring a great
deal. If one is piping a long log through \fBccze\fR, this option
might be turned off to speed up the process.

With the \fItransparent\fR option, one can make CCZE treat black
background colors as transparent \- that means, a black background
will appear transparent in a similar X terminal. If turned off, it
will appear as black.

If \fIcssfile\fR is set, then CCZE will not inline the Cascading Style
Sheet information into the outputted HTML, but include a link to the
external stylesheet given in this parameter.
.TP
.B \-p, \-\-plugin \fIPLUGIN\fR
While the default action is to load all plugins (see the \fBPlugins\fR
section below), when this option is present, only the specified
plugins will be loaded. If one knows what kind of log will be piped
through \fBccze\fR, using this option may result in a slight speedup.
.TP
.B \-r, \-\-remove\-facility
Syslog\-ng puts the facility level before log messages. With this
switch, these can be cut off.
.TP
.B \-\-help
Show summary of options and exit.
.TP
.B \-V, \-\-version
Show version of program.
.SH PLUGINS
Different programs have different kind of logs, and every kind of log
\fBccze\fR supports is implemented via a plug\-in. They are by default
located under \fI@pkglibdir@\fR and \fI$HOME/.ccze\fR (so they can
be overridden by the user easily).

At the moment, the following modules are bundled with the official
release:
.TP
.IR apm
For coloring apmd's logs.
.TP
.IR distcc
For coloring distccd's logs. (See \fIdistccd\fR(1))
.TP
.IR exim
For coloring exim's \fImain.log\fR. (See \fIexim\fR(8))
.TP
.IR fetchmail
For coloring fetchmail's log files. (See \fIfetchmail\fR(1))
.TP
.IR ftpstats
For coloring ftpStats compatible log files, such as Pure\-FTPD's
\fIStats\fR log format.
.TP
.IR httpd
For coloring apache\-style \fIaccess.log\fR and \fIerror.log\fR
files. (See \fIapache\fR(8))

As a side\-effect, all compatible formats, like \fIpure\-ftpd\fR(8)'s
common\-log format is also supported by this plugin.
.TP
.IR icecast
For coloring \fIicecast/icecast.log\fR and \fIicecast/usage.log\fR
files. (See \fIicecast\fR(8))
.TP
.IR oops
For coloring \fIoops/oops.log\fR files.
.TP
.IR php
For coloring \fIphp.log\fR files.
.TP
.IR postfix
For coloring postfix logs. (See \fIpostfix\fR(1))
.TP
.IR procmail
For coloring procmail's log file. (See \fIprocmail\fR(1))
.TP
.IR proftpd
For coloring proftpd's \fIaccess.log\fR and \fIauth.log\fR files. (See
\fIproftpd\fR(8))
.TP
.IR squid
For coloring squid's \fIaccess.log\fR, \fIstore.log\fR and
\fIcache.log\fR files. (See \fIsquid\fR(8))
.TP
.IR sulog
For coloring \fIsulog\fR files. (See \fIsu\fR(1))
.TP
.IR super
For coloring \fIsuper.log\fR files. (See \fIsuper\fR(1))
.TP
.IR syslog
For coloring generic syslog messages. (See \fIsyslogd\fR(8))
.TP
.IR ulogd
For coloring ulogd logs.
.TP
.IR vsftpd
For coloring \fIvsftpd.log\fR files. (See \fIvsftpd\fR(8))
.TP
.IR xferlog
For coloring \fIxferlog\fR files. (See \fIxferlog\fR(5))
.SH FILES
.TP
.IR @sysconfdir@/colorizerc ", " $HOME/.colorizerc
.RS
These files are the default configuration files for \fBcolorize\fR,
and are parsed by \fBccze\fR for the sake of full compatibility.
.RE
.IR @sysconfdir@/cczerc ", " $HOME/.cczerc
.RS
This two are the main configuration files, in which one can change the
colors used by the program to his liking. See the comments in the
beginning of \fI@sysconfdir@/cczerc\fR for a description on the files'
structure.

If neither of these files exist on your system, consider using the
\fBccze\-dump\fR utility in the source tree, which dumps the default
color set to standard output.
.RE
.SH "SEE ALSO"
.BR colorize (1)
.SH AUTHOR
ccze was written by Gergely Nagy <algernon@bonehunter.rulez.org>,
based on colorize by Istvan Karaszi <colorize@spam.raszi.hu>.


NEWS
NEWS -- list of user-visible changes		-*- outline -*-

* CCZE 0.2.1 [Fireflower] -- 2003-05-13

The Fireflower release brings transparency and background colour
support for the raw ANSI output mode, documentation updates and minor
bugfixes all over the place.

* CCZE 0.2.0 [Mephisto] -- 2003-03-29

Mephisto features raw ANSI output support, without the use of
nCurses. This adds support for scrollback, both on console and in
xterms. Also, it allows one to send CCZE to the background, and mix
various logs and options on one terminal.

A new coloriser for Pure-FTPD's Stats log format was added, and the
httpd module was modified to handle Pure-FTPD's CLF format, and bad
HTTP requests.

* CCZE 0.1.228 [Largo] -- 2003-03-20

This release features many many internal enhancements, mostly for the
plug-in system, a coloriser for proftpd logs, and various fixes all
over the place.

* CCZE 0.1.212 [Avatar] -- 2003-03-04

Implemented support for passing arguments to plugins. None of the
plugins shipped with CCZE support this, however, external plugins
might (and do, for that matter, see the homepage).

Also, some minor bugs (misformatting on the manual page, garbled
output, and not properly recognised procmail logs) were also squashed.

* CCZE 0.1.200 [Enigma of the Absolute] -- 2003-02-21

During the past few weeks, CCZE was ported to Cygwin and Tru64 (many
thanks to Jon Ewing and Erik Braun for their help).

Also, a coloriser for distccd.log files was added too.

* CCZE 0.1.171 [Mesmerism] -- 2003-02-09.

This release brings colorisers for php.log, oops/oops.log,
icecast/icecast.log and icecast/usage.log. Also, it is now possible to
disable any of the built-in plugins, that is, they behave as loadable
modules when --plugin is specified.

For the benefit of transparent xterm users, CCZE defaults to not alter
the background colour by default.

* CCZE 0.1.150 [The Cardinal Sin] -- 2003-02-04.

Since the last release, a great deal of code cleanup took place,
resulting in most of the modules' sources getting renamed. As usual,
bugs were fixed during this process, so certain characters will get
escaped in the HTML output.

Also, a full-blown coloriser module was added for xferlog files, and a
syslog-subcoloriser for postfix logs.

CCZE also handles SIGHUP nicely, by effectively restarting itself.

* CCZE 0.1.134 [Song of the Dispossessed] -- 2003-01-19.

Colours (even the background of the outputted HTML) can now be changed
on the command-line too. During this process, CCZE was ported to
FreeBSD.

* CCZE 0.1.122 [Wild in The Woods] -- 2003-01-13.

HTML output was cleaned up and made customisable. From now on, CCZE
uses Cascading Style Sheets.

* CCZE 0.1.108 [Devorzhum] -- 2003-01-10.

In this release experimental HTML output was added. It is not really
customisable yet, but produces usable output (better than that of
colorize! :).

* CCZE 0.1.103 [For My Fallen Angel] -- 2002.12.02.

Many internal cleanups and documentation updates, support for built-in
and partial plugins, and more.

* CCZE 0.1.91 [Like Gods of the Sun] -- 2002.11.19.

Major restructuring in the plugin code: cleaner, better, lighter. The
new design allows sub-colorisers, to colorise syslog messages (with
the date, host and that kind of stuff handled by the syslog module)
for example. Two such plugins are included: one for ulogd, another for
apmd.

* CCZE 0.1.80 [The Stance of Evander Sinque] -- 2002.10.12.

This is mainly a minor convenience-enhanced release. The major change
is that CCZE will now exit if it could not load any plugins.

* CCZE 0.1.72 [Base Level Erotica] -- 2002.10.04.

Brown paper-bag release, as I broke all the modules in 0.1.64. Apart
from this fix, this release is the same as 0.1.70 yesterday.

* CCZE 0.1.70 [Edenbeast] -- 2002.10.03.

Initial public release.


src/ccze.c
/* -*- mode: c; c-file-style: "gnu" -*-
 * ccze.c -- CCZE itself
 * Copyright (C) 2002, 2003 Gergely Nagy <algernon@bonehunter.rulez.org>
 *
 * This file is part of ccze.
 *
 * ccze is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * ccze is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
 * License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#include "system.h"

#ifdef HAVE_ARGP_H
# include <argp.h>
#endif
#include <ccze.h>
#include <dlfcn.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include "ccze-private.h"
#include "ccze-compat.h"

#define BLACK COLOR_PAIR (0)
#define RED COLOR_PAIR (1)
#define GREEN COLOR_PAIR (2)
#define YELLOW COLOR_PAIR (3)
#define BLUE COLOR_PAIR (4)
#define CYAN COLOR_PAIR (5)
#define MAGENTA COLOR_PAIR (6)
#define WHITE COLOR_PAIR (7)

#define ESC 0x1b

/* ccze somehow swaped cyan and magenta */
static int ccze_raw_ansi_color[] = {30, 31, 32, 33, 34, 36, 35, 37};

ccze_config_t ccze_config = {
  .scroll = 1,
  .convdate = 0,
  .remfac = 0,
  .wcol = 1,
  .slookup = 1,
  .rcfile = NULL,
  .cssfile = NULL,
  .transparent = 1,
  .pluginlist_len = 0,
  .pluginlist_alloc = 10,
  .color_argv_len = 0,
  .color_argv_alloc = 10,
  .mode = CCZE_MODE_CURSES
};

static short colors[] = {COLOR_BLACK, COLOR_RED, COLOR_GREEN, COLOR_YELLOW,
			 COLOR_BLUE, COLOR_CYAN, COLOR_MAGENTA, COLOR_WHITE};
static volatile sig_atomic_t sighup_received = 0;

#ifndef HAVE_ARGP_PARSE
const char *argp_program_name = "ccze";
#endif
const char *argp_program_version = "ccze 0.2." PATCHLEVEL;
const char *argp_program_bug_address = "<algernon@bonehunter.rulez.org>";
static struct argp_option options[] = {
  {"rcfile", 'F', "FILE", 0, "Read configuration from FILE", 1},
  {"html", 'h', NULL, 0, "Generate HTML output", 1},
  {"options", 'o', "OPTIONS...", 0, "Toggle some options\n"
   "(such as scroll, wordcolor and lookups, transparent, or cssfile)", 1},
  {"convert-date", 'C', NULL, 0, "Convert UNIX timestamps to readable format", 1},
  {"plugin", 'p', "PLUGIN", 0, "Load PLUGIN", 1},
  {"remove-facility", 'r', NULL, 0,
   "remove syslog-ng's facility from start of the lines", 1},
  {"color", 'c', "KEY=COLOR,...", 0, "Set the color of KEY to COLOR", 1},
  {"argument", 'a', "PLUGIN=ARGS...", 0, "Add ARGUMENTS to PLUGIN", 1},
  {"debug", 'd', NULL, OPTION_HIDDEN, "Turn on debugging.", 1},
  {"raw-ansi", 'A', NULL, 0, "Generate raw ANSI output", 1},
  {"list-plugins", 'l', NULL, 0, "List available plugins", 1},
  {"mode", 'm', "MODE", 0, "Change the output mode\n"
   "(Available modes are curses, ansi and html.)", 1},
  {NULL, 0, NULL, 0,  NULL, 0}
};
static error_t parse_opt (int key, char *arg, struct argp_state *state);
static struct argp argp =
  {options, parse_opt, 0, "ccze -- cheer up 'yer logs.", NULL, NULL, NULL};

enum
{
  CCZE_O_SUBOPT_SCROLL = 0,
  CCZE_O_SUBOPT_NOSCROLL,
  CCZE_O_SUBOPT_WORDCOLOR,
  CCZE_O_SUBOPT_NOWORDCOLOR,
  CCZE_O_SUBOPT_LOOKUPS,
  CCZE_O_SUBOPT_NOLOOKUPS,
  CCZE_O_SUBOPT_CSSFILE,
  CCZE_O_SUBOPT_NOCSSFILE,
  CCZE_O_SUBOPT_TRANSPARENT,
  CCZE_O_SUBOPT_NOTRANSPARENT,
  CCZE_O_SUBOPT_END
};

static char *o_subopts[] = {
  [CCZE_O_SUBOPT_SCROLL] = "scroll",
  [CCZE_O_SUBOPT_NOSCROLL] = "noscroll",
  [CCZE_O_SUBOPT_WORDCOLOR] = "wordcolor",
  [CCZE_O_SUBOPT_NOWORDCOLOR] = "nowordcolor",
  [CCZE_O_SUBOPT_LOOKUPS] = "lookups",
  [CCZE_O_SUBOPT_NOLOOKUPS] = "nolookups",
  [CCZE_O_SUBOPT_CSSFILE] = "cssfile",
  [CCZE_O_SUBOPT_NOCSSFILE] = "nocssfile",
  [CCZE_O_SUBOPT_TRANSPARENT] = "transparent",
  [CCZE_O_SUBOPT_NOTRANSPARENT] = "notransparent",
  [CCZE_O_SUBOPT_END] = NULL
};

enum
{
  CCZE_M_SUBOPT_CURSES,
  CCZE_M_SUBOPT_ANSI,
  CCZE_M_SUBOPT_HTML,
  CCZE_M_SUBOPT_DEBUG,
  CCZE_M_SUBOPT_END
};

static char *m_subopts[] = {
  [CCZE_M_SUBOPT_CURSES] = "curses",
  [CCZE_M_SUBOPT_ANSI] = "ansi",
  [CCZE_M_SUBOPT_HTML] = "html",
  [CCZE_M_SUBOPT_DEBUG] = "debug",
  [CCZE_M_SUBOPT_END] = NULL
};

static char *empty_subopts[] = { NULL };

static char *_strbrk_string;
static size_t _strbrk_string_len;

char *
ccze_strbrk (char *str, char delim)
{
  char *found;
      
  if (str)
    {
      _strbrk_string = str;
      _strbrk_string_len = strlen (str);
      found = str;
    }
  else
    found = _strbrk_string + 1;
  
  if (!_strbrk_string_len)
    return NULL;
  while (_strbrk_string_len >= 1 &&
	 *_strbrk_string != delim)
    {
      _strbrk_string++;
      _strbrk_string_len--;
    }
  if (_strbrk_string_len > 0)
    *_strbrk_string = '\0';
  return found;
}

char *
xstrdup (const char *str)
{
  if (!str)
    return NULL;
  else
    return strdup (str);
}

static error_t
parse_opt (int key, char *arg, struct argp_state *state)
{
  char *subopts, *value, *plugin;
  
  switch (key)
    {
    case 'c':
      subopts = arg;
      while (*subopts != '\0')
	{
	  ccze_getsubopt (&subopts, empty_subopts, &value);
	  ccze_config.color_argv[ccze_config.color_argv_len++] =
	    strdup (value);
	  if (ccze_config.color_argv_len >= ccze_config.color_argv_alloc)
	    {
	      ccze_config.color_argv_alloc *= 2;
	      ccze_config.color_argv =
		(char **)ccze_realloc (ccze_config.color_argv,
				       ccze_config.color_argv_alloc *
				       sizeof (char *));
	    }
	}
      break;
    case 'a':
      plugin = strtok (arg, "=");
      value = strtok (NULL, "\n");
      ccze_plugin_argv_set (plugin, value);
      break;
    case 'p':
      subopts = arg;
      while (*subopts != '\0')
	{
	  ccze_getsubopt (&subopts, empty_subopts, &value);
	  ccze_config.pluginlist[ccze_config.pluginlist_len++] =
	    strdup (value);
	  if (ccze_config.pluginlist_len >= ccze_config.pluginlist_alloc)
	    {
	      ccze_config.pluginlist_alloc *= 2;
	      ccze_config.pluginlist =
		(char **)ccze_realloc (ccze_config.pluginlist,
				       ccze_config.pluginlist_alloc *
				       sizeof (char *));
	    }
	}
      break;
    case 'h':
      ccze_config.mode = CCZE_MODE_HTML;
      break;
    case 'A':
      ccze_config.mode = CCZE_MODE_RAW_ANSI;
      break;
    case 'd':
      ccze_config.mode = CCZE_MODE_DEBUG;
      break;
    case 'l':
      ccze_config.mode = CCZE_MODE_PLUGIN_LIST;
      break;
    case 'F':
      ccze_config.rcfile = strdup (arg);
      break;
    case 'r':
      ccze_config.remfac = 1;
      break;
    case 'm':
      subopts = arg;
      while (*subopts != '\0')
	{
	  switch (getsubopt (&subopts, m_subopts, &value))
	    {
	    case CCZE_M_SUBOPT_CURSES:
	      ccze_config.mode = CCZE_MODE_CURSES;
	      break;
	    case CCZE_M_SUBOPT_ANSI:
	      ccze_config.mode = CCZE_MODE_RAW_ANSI;
	      break;
	    case CCZE_M_SUBOPT_HTML:
	      ccze_config.mode = CCZE_MODE_HTML;
	      break;
	    case CCZE_M_SUBOPT_DEBUG:
	      ccze_config.mode = CCZE_MODE_DEBUG;
	      break;
	    default:
	      argp_error (state, "unrecognised mode: `%s'", value);
	      break;
	    }
	}
      break;
    case 'o':
      subopts = arg;
      while (*subopts != '\0')
	{
	  switch (getsubopt (&subopts, o_subopts, &value))
	    {
	    case CCZE_O_SUBOPT_SCROLL:
	      ccze_config.scroll = 1;
	      break;
	    case CCZE_O_SUBOPT_NOSCROLL:
	      ccze_config.scroll = 0;
	      break;
	    case CCZE_O_SUBOPT_WORDCOLOR:
	      ccze_config.wcol = 1;
	      break;
	    case CCZE_O_SUBOPT_NOWORDCOLOR:
	      ccze_config.wcol = 0;
	      break;
	    case CCZE_O_SUBOPT_LOOKUPS:
	      ccze_config.slookup = 1;
	      break;
	    case CCZE_O_SUBOPT_NOLOOKUPS:
	      ccze_config.slookup = 0;
	      break;
	    case CCZE_O_SUBOPT_CSSFILE:
	      if (value)
		ccze_config.cssfile = strdup (value);
	      break;
	    case CCZE_O_SUBOPT_NOCSSFILE:
	      ccze_config.cssfile = NULL;
	      break;
	    case CCZE_O_SUBOPT_TRANSPARENT:
	      ccze_config.transparent = 1;
	      break;
	    case CCZE_O_SUBOPT_NOTRANSPARENT:
	      ccze_config.transparent = 0;
	      break;
	    default:
	      argp_error (state, "unrecognised option: `%s'", value);
	      break;
	    }
	}
      break;
    case 'C':
      ccze_config.convdate = 1;
      break;
    default:
      return ARGP_ERR_UNKNOWN;
    }
  return 0;
}

ccze_color_t
ccze_http_action (const char *method)
{
  if (!strcasecmp ("GET", method))
    return CCZE_COLOR_HTTP_GET;
  else if (!strcasecmp ("POST", method))
    return CCZE_COLOR_HTTP_POST;
  else if (!strcasecmp ("HEAD", method))
    return CCZE_COLOR_HTTP_HEAD;
  else if (!strcasecmp ("PUT", method))
    return CCZE_COLOR_HTTP_PUT;
  else if (!strcasecmp ("CONNECT", method))
    return CCZE_COLOR_HTTP_CONNECT;
  else if (!strcasecmp ("TRACE", method))
    return CCZE_COLOR_HTTP_TRACE;
  else
    return CCZE_COLOR_UNKNOWN;
}

void
ccze_print_date (const char *date)
{
  time_t ltime;
  char tmp[128];
  
  if (ccze_config.convdate)
    {
      ltime = atol (date);
      if (ltime < 0)
	{
	  ccze_addstr (CCZE_COLOR_DATE, date);
	  return;
	}
      strftime (tmp, sizeof (tmp) - 1, "%b %e %T", gmtime (&ltime));
      ccze_addstr (CCZE_COLOR_DATE, tmp);
    }
  else
    ccze_addstr (CCZE_COLOR_DATE, date);
}

void
ccze_newline (void)
{
  switch (ccze_config.mode)
    {
    case CCZE_MODE_HTML:
      printf ("<br>\n");
      break;
    case CCZE_MODE_DEBUG:
    case CCZE_MODE_RAW_ANSI:
      printf ("\n");
      break;
    case CCZE_MODE_CURSES:
      addstr ("\n");
      break;
    default:
      break;
    }
}

static char *
ccze_str_htmlencode (const char *src)
{
    char *buf, *dest;
    unsigned char c;

    dest = (char *)ccze_malloc (strlen (src) * 5 + 1);

    if (!dest)
      return NULL;

    buf = dest;
    while ((c = *src++))
      {
        switch (c)
	  {
	  case '>':
            *dest++ = '&';
            *dest++ = 'g';
            *dest++ = 't';
            *dest++ = ';';
            break;
	  case '<':
            *dest++ = '&';
            *dest++ = 'l';
            *dest++ = 't';
            *dest++ = ';';
            break;
	  case '&':
            *dest++ = '&';
            *dest++ = 'a';
            *dest++ = 'm';
            *dest++ = 'p';
            *dest++ = ';';
            break;
	  default:
            *dest++ = c;
	  }
      }
    *dest = '\0';
    return buf;
}

static void
ccze_addstr_internal (ccze_color_t col, const char *str, int enc)
{
  switch (ccze_config.mode)
    {
    case CCZE_MODE_HTML:
      if (str)
	{
	  char *d;

	  if (enc)
	    d = ccze_str_htmlencode (str);
	  else
	    d = strdup (str);
	  printf ("<font class=\"ccze_%s\">%s</font>",
		  ccze_color_lookup_name (col), d);
	  free (d);
	}
      break;
    case CCZE_MODE_RAW_ANSI:
      if (str)
	{
	  int c = ccze_color (col);

	  printf("%c[22m", ESC); /* default */

	  if (c & 0x1000) /* Bold */
	    {
	      printf("%c[1m", ESC);
	      c ^= 0x1000;
	    }
	  
	  if (c & 0x2000) /* Underline */
	    {
	      printf("%c[4m", ESC);
	      c ^= 0x2000;
	    }
	  
	  if (c & 0x4000) /* Reverse */
	    {
	      printf("%c[5m", ESC);
	      c ^= 0x4000;
	    }
	  
	  if (c & 0x8000) /* Blink */
	    {
	      printf("%c[7m", ESC);
	      c ^= 0x8000;
	    }

	  if (c >> 8 > 0 || !ccze_config.transparent)
	    printf("%c[%dm", ESC, ccze_raw_ansi_color[c >> 8] + 10);
	  printf ("%c[%dm%s%c[0m", ESC, ccze_raw_ansi_color[c & 0xf], str, ESC);
	}
      break;
    case CCZE_MODE_DEBUG:
      if (str)
	{
	  char *cn = ccze_color_lookup_name (col);
	  printf ("<%s>%s</%s>", cn, str, cn);
	}
      break;
    case CCZE_MODE_CURSES:
      attrset (ccze_color (col));
      addstr (str);
      break;
    default:
      break;
    }
}

void
ccze_addstr (ccze_color_t col, const char *str)
{
  ccze_addstr_internal (col, str, 1);
}

void
ccze_space (void)
{
  switch (ccze_config.mode)
    {
    case CCZE_MODE_HTML:
      ccze_addstr_internal (CCZE_COLOR_DEFAULT, "&nbsp;", 0);
      break;
    default:
      ccze_addstr (CCZE_COLOR_DEFAULT, " ");
      break;
    }
}

static void sigint_handler (int sig) __attribute__ ((noreturn));
static void
sigint_handler (int sig)
{
  switch (ccze_config.mode)
    {
    case CCZE_MODE_CURSES:
      endwin ();
      break;
    case CCZE_MODE_HTML:
      printf ("\n</body>\n</html>\n");
      break;
    case CCZE_MODE_RAW_ANSI:
      printf("%c[0m", ESC);
      break;
    default:
      break;
    }

  if (sig)
    {
      ccze_wordcolor_shutdown ();
      ccze_plugin_shutdown ();
    }
  
  exit (0);
}

static void
sigwinch_handler (int sig)
{
  endwin ();
  refresh ();
  signal (SIGWINCH, sigwinch_handler);
}

static void
sighup_handler (int sig)
{
  sighup_received = 1;
  signal (SIGHUP, sighup_handler);
}

static void
ccze_main (void)
{
  char *subject = NULL;
  size_t subjlen = 0;
  int i, j;
  char *homerc, *home;
  ccze_plugin_t **plugins;

  sighup_received = 0;
  ccze_color_init ();
  ccze_plugin_init ();

  if (ccze_config.rcfile)
    ccze_color_load (ccze_config.rcfile);
  else
    {
      ccze_color_load (SYSCONFDIR "/colorizerc");
      ccze_color_load (SYSCONFDIR "/cczerc");
      home = getenv ("HOME");
      if (home)
	{
	  asprintf (&homerc, "%s/.colorizerc", home);
	  ccze_color_load (homerc);
	  free (homerc);
	  asprintf (&homerc, "%s/.cczerc", home);
	  ccze_color_load (homerc);
	  free (homerc);
	}
    }

  while (ccze_config.color_argv_len > 0)
    {
      ccze_color_parse (ccze_config.color_argv[ccze_config.color_argv_len - 1]);
      free (ccze_config.color_argv[--ccze_config.color_argv_len]);
    }
  
  if (ccze_config.mode == CCZE_MODE_CURSES)
    {
      initscr ();
      signal (SIGWINCH, sigwinch_handler);
      nonl ();
      if (ccze_config.scroll)
	{
	  idlok (stdscr, TRUE);
	  scrollok (stdscr, TRUE);
	  leaveok (stdscr, FALSE);
	}
      
      start_color ();
      if (ccze_config.transparent)
	{
	  if (use_default_colors () == OK)
	    {
	      colors[0] = -1;
	    }
	}
      
      for (i = 0; i < 8; i++)
	for (j = 0; j < 8; j++)
	  init_pair (i*8 + j, colors[j], colors[i]);
    }
  else if (ccze_config.mode == CCZE_MODE_HTML)
    {
      printf
	("<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.01//Transitional//EN\">\n"
	 "<html>\n<head>\n<meta name=\"generator\" content=\"%s\">\n",
	 argp_program_version);

      if (ccze_config.cssfile)
	printf ("<link rel=\"stylesheet\" href=\"%s\">\n",
		ccze_config.cssfile);
      else
	{
	  printf ("<style type=\"text/css\">\n"
		  "body { font: 10pt courier; white-space: nowrap }\n");

	  ccze_colors_to_css ();

	  printf ("</style>\n");
	}
      
      
      printf ("<title>Log colorisation generated by %s</title>\n"
	      "</head>\n<body bgcolor=\"%s\">\n\n",
	      argp_program_version, ccze_cssbody_color ());
    }
  
  signal (SIGINT, sigint_handler);
  signal (SIGHUP, sighup_handler);
  
  ccze_wordcolor_setup ();

  if (ccze_config.pluginlist_len == 0)
    ccze_plugin_load_all ();
  else
    {
      int pl = ccze_config.pluginlist_len;
      while (pl-- > 0)
	ccze_plugin_load (ccze_config.pluginlist[pl]);
    }
  ccze_plugin_load_all_builtins ();
  ccze_plugin_finalise ();
      
  i = 0;
  plugins = ccze_plugins ();
  if (!plugins[0])
    {
      endwin ();
      fprintf (stderr, "ccze: No plugins found. Exiting.\n");
      exit (1);
    }

  ccze_plugin_argv_finalise ();

  if (ccze_config.mode == CCZE_MODE_PLUGIN_LIST)
    {
      ccze_plugin_list_fancy ();
      sigint_handler (0);
    }
  
  ccze_plugin_setup ();
  
  while ((getline (&subject, &subjlen, stdin) != -1) && !sighup_received)
    {
      int handled = 0;
      int status = 0;
      char *rest = NULL, *rest2 = NULL;
      char *tmp = strchr (subject, '\n');
      unsigned int remfac_tmp;

      if (tmp)
	tmp[0] = '\0';

      if (ccze_config.remfac && (sscanf (subject, "<%u>", &remfac_tmp) > 0))
	{
	  tmp = strdup (strchr (subject, '>') + 1);
	  free (subject);
	  subject = tmp;
	}

      subjlen = strlen (subject);
      ccze_plugin_run (plugins, subject, subjlen, &rest,
		       CCZE_PLUGIN_TYPE_FULL, &handled, &status);
      
      if (rest)
	{
	  handled = 0;
	  ccze_plugin_run (plugins, rest, strlen (rest), &rest2,
			   CCZE_PLUGIN_TYPE_PARTIAL, &handled, &status);
	  if (handled == 0)
	    ccze_wordcolor_process (rest, ccze_config.wcol,
				    ccze_config.slookup);
	  else
	    ccze_wordcolor_process (rest2, ccze_config.wcol,
				    ccze_config.slookup);
	  ccze_newline ();
	  free (rest);
	  free (rest2);
	}

      if (status == 0)
	{
	  ccze_wordcolor_process (subject, ccze_config.wcol,
				  ccze_config.slookup);
	  ccze_newline ();
	}

      if (ccze_config.mode == CCZE_MODE_CURSES)
	refresh ();
    }

  if (ccze_config.mode == CCZE_MODE_CURSES)
    refresh ();
}

int
main (int argc, char **argv)
{
      
  ccze_config.pluginlist = (char **)ccze_calloc (ccze_config.pluginlist_alloc,
						 sizeof (char *));
  ccze_config.color_argv = (char **)ccze_calloc (ccze_config.color_argv_alloc,
						 sizeof (char *));
  ccze_plugin_argv_init ();
  argp_parse (&argp, argc, argv, 0, 0, NULL);

  do
    {
      ccze_main ();
      ccze_wordcolor_shutdown ();
      ccze_plugin_shutdown ();
    } while (sighup_received);

  sigint_handler (0);
  
  return 0;
}
